### PR TITLE
PIMS-1685 Adding code coverage reports

### PR DESCRIPTION
**Pull request by Fosol**
## Description

Updating GitHub Actions to generate code coverage files and send them to CodeCov.

Updating documentation to help new developers.

### CODE REVIEW

File: .github/workflows/api-dotnetcore.yml [hunk 1]
```diff
@@ -21,9 +21,34 @@ jobs:
...
+        run: coverlet ./tests/unit/dal/bin/Release/netcoreapp3.1/Pims.Dal.Test.dll --target "dotnet" --targetargs "test ./ --no-build" -o "./tests/TestResults/coverage.json" --exclude "[*.Test]*" --exclude "[*]*Model" --exclude-by-attribute "CompilerGenerated" --merge-with "./tests/TestResults/coverage.json" -f json
+        working-directory: ${{env.working-directory}}
+
+      - name: Codecov
+        uses: codecov/codecov-action@v1.0.7
+        with:
+          # User defined upload name. Visible in Codecov UI
+          name: PIMS
+          # Repository upload token - get it from codecov.io. Required only for private repositories
+          token: c02e6ccf-6e0f-47cd-842e-ab2f488933b8
```
Comment (Fosol) [inline .github/workflows/api-dotnetcore.yml:46]: At some point I'll need to regenerate this token and find a way to provide it outside of source.

File: .github/workflows/api-dotnetcore.yml [hunk 2]
```diff
@@ -21,9 +21,34 @@ jobs:
...
+        run: dotnet tool install -g coverlet.console
+        working-directory: ${{env.working-directory}}
       - name: Build
         run: dotnet build --configuration Release --no-restore
         working-directory: ${{env.working-directory}}
       - name: Test
         run: dotnet test --no-restore --verbosity normal
         working-directory: ${{env.working-directory}}
+      - name: Generate code coverage
+        run: coverlet ./tests/unit/api/bin/Release/netcoreapp3.1/Pims.Api.Test.dll --target "dotnet" --targetargs "test ./ --no-build" -o "./tests/TestResults/coverage.json" --exclude "[*.Test]*" --exclude "[*]*Model" --exclude-by-attribute "CompilerGenerated" -f json
```
Comment (Fosol) [inline .github/workflows/api-dotnetcore.yml:34]: Generate code coverage files and merge them to send to CodeCov.

File: .github/workflows/app-react.yml [hunk 3]
```diff
@@ -27,5 +27,21 @@ jobs:
...
+      - run: npm run coverage
         working-directory: ${{env.working-directory}}
+
+      - name: Codecov
+        uses: codecov/codecov-action@v1.0.7
+        with:
+          # User defined upload name. Visible in Codecov UI
+          name: PIMS
+          # Repository upload token - get it from codecov.io. Required only for private repositories
+          token: c02e6ccf-6e0f-47cd-842e-ab2f488933b8
```
Comment (Fosol) [inline .github/workflows/app-react.yml:39]: At some point I'll need to regenerate this token and find a way to provide it outside of source.

File: README.md [hunk 4]
```diff
@@ -1,6 +1,7 @@
 [![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](LICENSE)
-![API (DEV)](https://github.com/bcgov/PIMS/workflows/API%20(.NET%20Core)/badge.svg?branch=master)
-![APP (React)](https://github.com/bcgov/PIMS/workflows/APP%20(React)/badge.svg?branch=master)
+![API (DEV)](<https://github.com/bcgov/PIMS/workflows/API%20(.NET%20Core)/badge.svg?branch=master>)
+![APP (React)](<https://github.com/bcgov/PIMS/workflows/APP%20(React)/badge.svg?branch=master>)
+[![codecov](https://codecov.io/gh/bcgov/PIMS/branch/master/graph/badge.svg)](https://codecov.io/gh/bcgov/PIMS)
```
Comment (Fosol) [inline README.md:4]: Master branch codecov badge.

File: README.md [hunk 5]
```diff
@@ -49,13 +50,16 @@ Objectives include:
...
-![APP (React)](https://github.com/bcgov/PIMS/workflows/APP%20(React)/badge.svg?branch=dev)
-
-**IN DEVELOPMENT**
+> **IN DEVELOPMENT**
 
 ## Development
 
+![API (.Net Core)](<https://github.com/bcgov/PIMS/workflows/API%20(.NET%20Core)/badge.svg?branch=dev>)
+![APP (React)](<https://github.com/bcgov/PIMS/workflows/APP%20(React)/badge.svg?branch=dev>)
+[![codecov](https://codecov.io/gh/bcgov/PIMS/branch/dev/graph/badge.svg)](https://codecov.io/gh/bcgov/PIMS)
```
Comment (Fosol) [inline README.md:59]: Dev branch codecov badge.

File: codecov.yml [hunk 6]
```diff
@@ -0,0 +1,21 @@
+codecov:
```
Comment (Fosol) [inline codecov.yml:1]: A way to configure CodeCov.

File: docs/DEVELOPMENT.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# Development
```
Comment (Fosol) [inline docs/DEVELOPMENT.md:1]: Help new developers get started by putting all the information in one place.

File: frontend/package.json [hunk 8]
```diff
@@ -94,6 +94,7 @@
     "build": "cross-env CI=true react-scripts build",
     "test": "cross-env CI=true react-scripts test --env=jsdom-fourteen --verbose --watchAll=false --runInBand",
     "test:watch": "react-scripts test --env=jsdom-fourteen",
+    "coverage": "cross-env CI=true react-scripts test --coverage --env=jsdom-fourteen --watchAll=false --runInBand",
```
Comment (Fosol) [inline frontend/package.json:97]: Generate code coverage

File: frontend/package.json [hunk 9]
```diff
@@ -102,13 +103,31 @@
     "check": "prettier --check \"./src/**/*.{js,jsx,ts,tsx,css,scss}\"",
     "pre-push:check": "npm run check && npm run lint"
   },
+  "jest": {
```
Comment (Fosol) [inline frontend/package.json:106]: Configure code coverage

File: scripts/gen-env-files.sh [hunk 10]
```diff
@@ -25,7 +25,7 @@ DB_USER=$varKeycloakDb
 DB_PASSWORD=$passvar
 KEYCLOAK_USER=$varKeycloak
 KEYCLOAK_PASSWORD=$passvar
-KEYCLOAK_IMPORT=/tmp/realm-export.json -Dkeycloak.profile.feature.scripts=enabled -Dkeycloak.profile.feature.upload_scripts=enabled
+# KEYCLOAK_IMPORT=/tmp/realm-export.json -Dkeycloak.profile.feature.scripts=enabled -Dkeycloak.profile.feature.upload_scripts=enabled
```
Comment (Fosol) [inline scripts/gen-env-files.sh:28]: This no longer works, but I've left it in for documentation purposes.

File: .github/workflows/app-react.yml (hunk 3 above)
Comment (asanchezr) [reply .github/workflows/app-react.yml:39]: yeah, this got already merged to source control so nothing we can do now but I believe GitHub provides a way to create secrets?

Comment (asanchezr) [reply .github/workflows/app-react.yml:39]: https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets

and

https://help.github.com/en/actions/configuring-and-managing-workflows/using-variables-and-secrets-in-a-workflow

**Pull request merged**